tachyon/math/base/byinverter.h

```diff
@@ -0,0 +1,469 @@
+      usize zeros = trailing0(g);
+      zeros = least(zeros, steps);
+      steps -= zeros;
+      delta += zeros;
+      g >>= zeros;
+      t[0][0] <<= zeros;
+      t[0][1] <<= zeros;
+
+      if (steps == 0) {
+        break;
+      }
+
+      if (delta > 0) {
+        delta = -delta;
+
+        y[0] = -t[0][0];
+        y[1] = -t[0][1];
+        t[0][0] = t[1][0];
+        t[0][1] = t[1][1];
+        t[1][0] = y[0];
+        t[1][1] = y[1];
+
+        x = -f;
+        f = g;
+        g = x;
+      }
+
+      // The formula (3 * x) xor 28 = -1 / x (mod 32) for an odd integer
+      // x in the two's complement code has been derived from the formula
+      // (3 * x) xor 2 = 1 / x (mod 32) attributed to Peter Montgomery
+      u64 mask = (1 << least(least(steps, 1 - delta), 5)) - 1;
+      u64 w = g * ((3 * f) ^ 28) & mask;
+
+      y[0] = t[0][0] * w + t[1][0];
+      y[1] = t[0][1] * w + t[1][1];
+      t[1][0] = y[0];
+      t[1][1] = y[1];
+
+      g += w * f;
+    }
+  }
+
+  /// Stores the updated values of the variables f and g for specified
+  /// initial ones and Bernstein-Yang transition matrix multiplied by 2^62.
+  /// In the vector form this operation can be described using the formula:
+  /// "(f, g)' := matrix * (f, g)' / 2^62", where "'" is the transpose
+  /// operator and ":=" denotes assignment
+  static void fg(LCInt &f, LCInt &g, const i64 t[2][2]) {
+    LCInt x = (f * t[0][0] + g * t[0][1]).shift();
+    LCInt y = (f * t[1][0] + g * t[1][1]).shift();
+    f = x;
+    g = y;
+  }
+
+  /// Stores the updated values of the variables d and e for specified
+  /// initial ones and Bernstein-Yang transition matrix multiplied by
+  /// 2^62. The new value of the vector "(d, e)'" is congruent modulo M
+  /// to "matrix * (d, e)' / 2^62 (mod M)", where M is the modulus the
+  /// inverter was created for and "'" stands for the transpose operator.
+  /// Both the initial and new values of d and e lie in (-2 * M, M)
+  void de(LCInt &d, LCInt &e, const i64 (&t)[2][2]) const {
+    i64 md = t[0][0] * d.is_negative() + t[0][1] * e.is_negative();
+    i64 me = t[1][0] * d.is_negative() + t[1][1] * e.is_negative();
+    {
+      i64 cd = (t[0][0] * d.lowest() + t[0][1] * e.lowest()) & LCInt::MASK;
+      i64 ce = (t[1][0] * d.lowest() + t[1][1] * e.lowest()) & LCInt::MASK;
+
+      md -= (cd * this->inverse62 + md) & LCInt::MASK;
+      me -= (ce * this->inverse62 + me) & LCInt::MASK;
+    }
+    LCInt cd = d * t[0][0] + e * t[0][1] + this->modulus * md;
+    LCInt ce = d * t[1][0] + e * t[1][1] + this->modulus * me;
+    d = cd.shift();
+    e = ce.shift();
+  }
+
+  /// Stores in the variable "value" either "value (mod M)" or
+  /// "-value (mod M)", where M is the modulus the inverter was
+  /// created for, depending on "negate", which determines the
+  /// presence of "-" in the used formula. The initial integer
+  /// contained in "value" lies in (-2 * M, M)
+  void norm(LCInt &value, const bool negate) const {
```

Comment:
Please make the function return the result of the function, e.g `Inverse(x) -> x⁻¹`